src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy

```diff
@@ -1720,45 +1796,339 @@ class ActionDelegate {
         return menuItem
     }
 
-    protected Case doCreateMenuItem(String uri, String identifier, Case filter, Case orgGroup, Map<String, I18nString> allowedRoles, Map<String, I18nString> bannedRoles, List<String> defaultHeaders) {
-        if (findMenuItem(identifier)) {
-            throw new IllegalArgumentException("Menu item identifier $identifier is not unique!")
+    Case createMenuItem(MenuItemBody body) {
+        String sanitizedIdentifier = sanitize(body.identifier)
+
+        if (findMenuItem(sanitizedIdentifier)) {
+            throw new IllegalArgumentException("Menu item identifier $sanitizedIdentifier is not unique!")
         }
-        orgGroup = orgGroup ?: nextGroupService.findDefaultGroup()
-        Case itemCase = createCase(FilterRunner.PREFERRED_FILTER_ITEM_NET_IDENTIFIER, filter.title)
-        itemCase.setUriNodeId(uriService.findByUri(uri).id)
-        itemCase.dataSet[PREFERENCE_ITEM_FIELD_ALLOWED_ROLES].options = allowedRoles
-        itemCase.dataSet[PREFERENCE_ITEM_FIELD_BANNED_ROLES].options = bannedRoles
-        itemCase = workflowService.save(itemCase)
-        Task newItemTask = findTask { it._id.eq(new ObjectId(itemCase.tasks.find { it.transition == "init" }.task)) }
+
+        Case parentItemCase = getOrCreateFolderItem(body.uri)
+
+        Case menuItemCase = createCase(FilterRunner.PREFERRED_ITEM_NET_IDENTIFIER, body.filter.title)
+        menuItemCase.setUriNodeId(uriService.findByUri(body.uri).id)
+        menuItemCase.dataSet[PREFERENCE_ITEM_FIELD_ALLOWED_ROLES].options = body.allowedRoles
+        menuItemCase.dataSet[PREFERENCE_ITEM_FIELD_BANNED_ROLES].options = body.bannedRoles
+        if (parentItemCase != null) {
+            parentItemCase = appendChildCaseIdAndSave(parentItemCase, menuItemCase.stringId)
+        }
+        menuItemCase = workflowService.save(menuItemCase)
+        Task newItemTask = findTask { it._id.eq(new ObjectId(menuItemCase.tasks.find { it.transition == "initialize" }.task)) }
+        String nodePath = createNodePath(body.uri, sanitizedIdentifier)
+        uriService.getOrCreate(nodePath, UriContentType.CASE)
+
         assignTask(newItemTask)
+        I18nString newName = body.name ? body.name : (body.filter.dataSet[FILTER_FIELD_I18N_FILTER_NAME].value as I18nString)
```

Comment:
Please use [Elvis operator](https://groovy-lang.org/operators.html#_elvis_operator) for better readability 
```groovy
body.name ?: (body.filter.dataSet[FILTER_FIELD_I18N_FILTER_NAME].value as I18nString)
```